docs/proposals/image-index.md

```diff
@@ -0,0 +1,153 @@
+  * For that I create a new classification group named “os” and provide an “selector” that matches the labels
+    in the target “classification” objects.
+
+
+### Proposed API Changes
+
+For the implementation, two new API types are needed:
+
+```go
+// ImageClass provides a single Image "class" a context that a
+// certain group if Images shares. For example "rhel7" or "python".
+// To create an ImageClass, you have to point it to an initial
+// or "target" image first. From that point, the ImageClassController
+// will track the history of that Image.
+type ImageClass struct{
+		unversioned.TypeMeta
+		kapi.ObjectMeta
+
+		Current      ImageChain   
+		History      []ImageChain 
+}
+
+// ImageClassGroup groups the various ImageClass objects into a
+// logical group. For example a "vendor" group groups all ImageClass
+// objects that matches the defined selector.
+type ImageClassGroup struct {
+		unversioned.TypeMeta
+		kapi.ObjectMeta
+
+		Selector: map[string]string
+}
+
+// ImageChain represents an internal state of the Image layers and
+// the chains based on the image layers.
+type ImageChain struct {
+		Chains    []string
+		LastChain string
+}
+```
+
+The `ImageClass` object is needed to keep the historical informations about the
+image persistent as the in-memory image chain index will only contain the
+images and chains that are currently available in the cluster.
+
+In future this allows to see which images are based on old or insecure images
+and mark the affected images as outdated.
+
+
+### Image Chains Index
+
+The Image is represented by the list of its layers. To be able to identify the
+base image, we have to introspect the layers and calculate the "hash chains"
+for the image layers. We also need to keep this index in memory for fast
+access.
+
+For example:
+
+* The `base` image contains 3 layers: `AAA`, `BBB` and `CCC`
+  * For that we have 3 chains: `chain1(AAA)`, `chain2(AAA, BBB)`, `chain3(AAA, BBB, CCC)`
+
+* An `ancestor` image contains 4 layers: `AAA`, `BBB`, `CCC` and `DDD`
+  * To say that the "ancestor" image is based on the `base` image, we can
+    assume that the "ancestor" image must contain the `chain3` in its chains list.
+
+
+For the implementation we need a indexed informer that contains two indexes:
+
+* `ChainIndex` that provides mapping between a chain and the list of images containing
+  this chain:
+  * `chain => [image1, image2, image3]`
+* `LastChainIndex` that provides mapping between "last" chain and the image that it represents:
+  * `chain => image1`
+
+The memory footprint for this index can be estimated by the size of the items
+that have to be stored. The `ChainIndex` needs a single key for each image layer,
+and the chain itself is represented by the 64 strings (SHA256). The Images
+re-use layers so the cost goes down with the number of re-used layers.
+
+The `LastChainIndex` cost is the same as the number of images in the cluster. The
+key represents an pointer to the image object.
+
```

Comment:
I know what I asked you to add and forgot about it - the CLI changes, we're planning to add. I mean `oadm classify` and how it will work.